From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcrt: Implement details::_Context's constructors and _StackGuard::_GetCurrentInlineDepth.

details::_Context's constructor from a Context pointer and its default
constructor closure (no context), _Context::_Oversubscribe (the same
function as Context::Oversubscribe, still a FIXME), and
_StackGuard::_GetCurrentInlineDepth, a counter of the current thread's
context the _StackGuard objects in the headers count inline chores with.
msvcr110, msvcr120 (msvcr120_app forwards) and concrt140.
---
 dlls/concrt140/concrt140.spec       | 18 +++++++--------
 dlls/msvcr110/msvcr110.spec         | 18 +++++++--------
 dlls/msvcr120/msvcr120.spec         | 18 +++++++--------
 dlls/msvcr120_app/msvcr120_app.spec | 18 +++++++--------
 dlls/msvcrt/concurrency.c           | 45 +++++++++++++++++++++++++++++++++++++
 5 files changed, 81 insertions(+), 36 deletions(-)

diff --git a/dlls/concrt140/concrt140.spec b/dlls/concrt140/concrt140.spec
index c7cb572..a34b11d 100644
--- a/dlls/concrt140/concrt140.spec
+++ b/dlls/concrt140/concrt140.spec
@@ -24,9 +24,9 @@
 @ cdecl -arch=arm ??0_Condition_variable@details@Concurrency@@QAA@XZ(ptr) _Condition_variable_ctor
 @ thiscall -arch=i386 ??0_Condition_variable@details@Concurrency@@QAE@XZ(ptr) _Condition_variable_ctor
 @ cdecl -arch=win64 ??0_Condition_variable@details@Concurrency@@QEAA@XZ(ptr) _Condition_variable_ctor
-@ stub -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z
-@ stub -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z
-@ stub -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z
+@ cdecl -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z(ptr ptr) _Context_ctor
+@ thiscall -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z(ptr ptr) _Context_ctor
+@ cdecl -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z(ptr ptr) _Context_ctor
 @ cdecl -arch=arm ??0_NonReentrantBlockingLock@details@Concurrency@@QAA@XZ(ptr) _ReentrantBlockingLock_ctor
 @ thiscall -arch=i386 ??0_NonReentrantBlockingLock@details@Concurrency@@QAE@XZ(ptr) _ReentrantBlockingLock_ctor
 @ cdecl -arch=win64 ??0_NonReentrantBlockingLock@details@Concurrency@@QEAA@XZ(ptr) _ReentrantBlockingLock_ctor
@@ -318,9 +318,9 @@
 @ cdecl -arch=arm ??_F?$_SpinWait@$0A@@details@Concurrency@@QAAXXZ(ptr) SpinWait_dtor
 @ thiscall -arch=i386 ??_F?$_SpinWait@$0A@@details@Concurrency@@QAEXXZ(ptr) SpinWait_dtor
 @ cdecl -arch=win64 ??_F?$_SpinWait@$0A@@details@Concurrency@@QEAAXXZ(ptr) SpinWait_dtor
-@ stub -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ
-@ stub -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ
-@ stub -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ
+@ cdecl -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ(ptr) _Context_default_ctor
+@ thiscall -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ(ptr) _Context_default_ctor
+@ cdecl -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ(ptr) _Context_default_ctor
 @ cdecl -arch=arm ??_F_Scheduler@details@Concurrency@@QAAXXZ(ptr) _Scheduler_ctor
 @ thiscall -arch=i386 ??_F_Scheduler@details@Concurrency@@QAEXXZ(ptr) _Scheduler_ctor
 @ cdecl -arch=win64 ??_F_Scheduler@details@Concurrency@@QEAAXXZ(ptr) _Scheduler_ctor
@@ -470,8 +470,8 @@
 @ stub -arch=win32 ?_GetConcRTTraceInfo@Concurrency@@YAPBU_CONCRT_TRACE_INFO@details@1@XZ
 @ stub -arch=win64 ?_GetConcRTTraceInfo@Concurrency@@YAPEBU_CONCRT_TRACE_INFO@details@1@XZ
 @ cdecl ?_GetConcurrency@details@Concurrency@@YAIXZ() _GetConcurrency
-@ stub -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ
-@ stub -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ
+@ cdecl -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ() _StackGuard__GetCurrentInlineDepth
+@ cdecl -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ() _StackGuard__GetCurrentInlineDepth
 @ cdecl ?_GetNumberOfVirtualProcessors@_CurrentScheduler@details@Concurrency@@SAIXZ() _CurrentScheduler__GetNumberOfVirtualProcessors
 @ cdecl -arch=arm ?_GetScheduler@_Scheduler@details@Concurrency@@QAAPAVScheduler@3@XZ(ptr) _Scheduler__GetScheduler
 @ thiscall -arch=i386 ?_GetScheduler@_Scheduler@details@Concurrency@@QAEPAVScheduler@3@XZ(ptr) _Scheduler__GetScheduler
@@ -554,7 +554,7 @@
 @ cdecl -arch=arm ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAAKXZ(ptr) SpinWait__NumberOfSpins
 @ thiscall -arch=i386 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAEKXZ(ptr) SpinWait__NumberOfSpins
 @ cdecl -arch=win64 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IEAAKXZ(ptr) SpinWait__NumberOfSpins
-@ stub ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z
+@ cdecl ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z(long) Context_Oversubscribe
 @ cdecl -arch=arm ?_Reference@_Scheduler@details@Concurrency@@QAAIXZ(ptr) _Scheduler__Reference
 @ thiscall -arch=i386 ?_Reference@_Scheduler@details@Concurrency@@QAEIXZ(ptr) _Scheduler__Reference
 @ cdecl -arch=win64 ?_Reference@_Scheduler@details@Concurrency@@QEAAIXZ(ptr) _Scheduler__Reference
diff --git a/dlls/msvcr110/msvcr110.spec b/dlls/msvcr110/msvcr110.spec
index 663fb9d..e584e4a 100644
--- a/dlls/msvcr110/msvcr110.spec
+++ b/dlls/msvcr110/msvcr110.spec
@@ -23,9 +23,9 @@
 @ cdecl -arch=arm ??0_Condition_variable@details@Concurrency@@QAA@XZ(ptr) _Condition_variable_ctor
 @ thiscall -arch=i386 ??0_Condition_variable@details@Concurrency@@QAE@XZ(ptr) _Condition_variable_ctor
 @ cdecl -arch=win64 ??0_Condition_variable@details@Concurrency@@QEAA@XZ(ptr) _Condition_variable_ctor
-@ stub -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z
-@ stub -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z
-@ stub -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z
+@ cdecl -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z(ptr ptr) _Context_ctor
+@ thiscall -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z(ptr ptr) _Context_ctor
+@ cdecl -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z(ptr ptr) _Context_ctor
 @ cdecl -arch=arm ??0_Interruption_exception@details@Concurrency@@QAA@PBD@Z(ptr str) _Interruption_exception_ctor_str
 @ thiscall -arch=i386 ??0_Interruption_exception@details@Concurrency@@QAE@PBD@Z(ptr str) _Interruption_exception_ctor_str
 @ cdecl -arch=win64 ??0_Interruption_exception@details@Concurrency@@QEAA@PEBD@Z(ptr str) _Interruption_exception_ctor_str
@@ -380,9 +380,9 @@
 @ cdecl -arch=arm ??_F?$_SpinWait@$0A@@details@Concurrency@@QAAXXZ(ptr) SpinWait_dtor
 @ thiscall -arch=i386 ??_F?$_SpinWait@$0A@@details@Concurrency@@QAEXXZ(ptr) SpinWait_dtor
 @ cdecl -arch=win64 ??_F?$_SpinWait@$0A@@details@Concurrency@@QEAAXXZ(ptr) SpinWait_dtor
-@ stub -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ
-@ stub -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ
-@ stub -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ
+@ cdecl -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ(ptr) _Context_default_ctor
+@ thiscall -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ(ptr) _Context_default_ctor
+@ cdecl -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ(ptr) _Context_default_ctor
 @ cdecl -arch=arm ??_F_Scheduler@details@Concurrency@@QAAXXZ(ptr) _Scheduler_ctor
 @ thiscall -arch=i386 ??_F_Scheduler@details@Concurrency@@QAEXXZ(ptr) _Scheduler_ctor
 @ cdecl -arch=win64 ??_F_Scheduler@details@Concurrency@@QEAAXXZ(ptr) _Scheduler_ctor
@@ -534,8 +534,8 @@
 @ stub -arch=win32 ?_GetConcRTTraceInfo@Concurrency@@YAPBU_CONCRT_TRACE_INFO@details@1@XZ
 @ stub -arch=win64 ?_GetConcRTTraceInfo@Concurrency@@YAPEBU_CONCRT_TRACE_INFO@details@1@XZ
 @ cdecl ?_GetConcurrency@details@Concurrency@@YAIXZ() _GetConcurrency
-@ stub -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ
-@ stub -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ
+@ cdecl -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ() _StackGuard__GetCurrentInlineDepth
+@ cdecl -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ() _StackGuard__GetCurrentInlineDepth
 @ cdecl ?_GetNumberOfVirtualProcessors@_CurrentScheduler@details@Concurrency@@SAIXZ() _CurrentScheduler__GetNumberOfVirtualProcessors
 @ cdecl -arch=arm ?_GetScheduler@_Scheduler@details@Concurrency@@QAAPAVScheduler@3@XZ(ptr) _Scheduler__GetScheduler
 @ thiscall -arch=i386 ?_GetScheduler@_Scheduler@details@Concurrency@@QAEPAVScheduler@3@XZ(ptr) _Scheduler__GetScheduler
@@ -567,7 +567,7 @@
 @ cdecl -arch=arm ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAAKXZ(ptr) SpinWait__NumberOfSpins
 @ thiscall -arch=i386 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAEKXZ(ptr) SpinWait__NumberOfSpins
 @ cdecl -arch=win64 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IEAAKXZ(ptr) SpinWait__NumberOfSpins
-@ stub ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z
+@ cdecl ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z(long) Context_Oversubscribe
 @ cdecl -arch=arm ?_Reference@_Scheduler@details@Concurrency@@QAAIXZ(ptr) _Scheduler__Reference
 @ thiscall -arch=i386 ?_Reference@_Scheduler@details@Concurrency@@QAEIXZ(ptr) _Scheduler__Reference
 @ cdecl -arch=win64 ?_Reference@_Scheduler@details@Concurrency@@QEAAIXZ(ptr) _Scheduler__Reference
diff --git a/dlls/msvcr120/msvcr120.spec b/dlls/msvcr120/msvcr120.spec
index 69800cd..1d7599b 100644
--- a/dlls/msvcr120/msvcr120.spec
+++ b/dlls/msvcr120/msvcr120.spec
@@ -20,9 +20,9 @@
 @ cdecl -arch=arm ??0_Condition_variable@details@Concurrency@@QAA@XZ(ptr) _Condition_variable_ctor
 @ thiscall -arch=i386 ??0_Condition_variable@details@Concurrency@@QAE@XZ(ptr) _Condition_variable_ctor
 @ cdecl -arch=win64 ??0_Condition_variable@details@Concurrency@@QEAA@XZ(ptr) _Condition_variable_ctor
-@ stub -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z
-@ stub -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z
-@ stub -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z
+@ cdecl -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z(ptr ptr) _Context_ctor
+@ thiscall -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z(ptr ptr) _Context_ctor
+@ cdecl -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z(ptr ptr) _Context_ctor
 @ cdecl -arch=arm ??0_Interruption_exception@details@Concurrency@@QAA@PBD@Z(ptr str) _Interruption_exception_ctor_str
 @ thiscall -arch=i386 ??0_Interruption_exception@details@Concurrency@@QAE@PBD@Z(ptr str) _Interruption_exception_ctor_str
 @ cdecl -arch=win64 ??0_Interruption_exception@details@Concurrency@@QEAA@PEBD@Z(ptr str) _Interruption_exception_ctor_str
@@ -377,9 +377,9 @@
 @ cdecl -arch=arm ??_F?$_SpinWait@$0A@@details@Concurrency@@QAAXXZ(ptr) SpinWait_dtor
 @ thiscall -arch=i386 ??_F?$_SpinWait@$0A@@details@Concurrency@@QAEXXZ(ptr) SpinWait_dtor
 @ cdecl -arch=win64 ??_F?$_SpinWait@$0A@@details@Concurrency@@QEAAXXZ(ptr) SpinWait_dtor
-@ stub -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ
-@ stub -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ
-@ stub -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ
+@ cdecl -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ(ptr) _Context_default_ctor
+@ thiscall -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ(ptr) _Context_default_ctor
+@ cdecl -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ(ptr) _Context_default_ctor
 @ cdecl -arch=arm ??_F_Scheduler@details@Concurrency@@QAAXXZ(ptr) _Scheduler_ctor
 @ thiscall -arch=i386 ??_F_Scheduler@details@Concurrency@@QAEXXZ(ptr) _Scheduler_ctor
 @ cdecl -arch=win64 ??_F_Scheduler@details@Concurrency@@QEAAXXZ(ptr) _Scheduler_ctor
@@ -528,8 +528,8 @@
 @ stub -arch=win32 ?_GetConcRTTraceInfo@Concurrency@@YAPBU_CONCRT_TRACE_INFO@details@1@XZ
 @ stub -arch=win64 ?_GetConcRTTraceInfo@Concurrency@@YAPEBU_CONCRT_TRACE_INFO@details@1@XZ
 @ cdecl ?_GetConcurrency@details@Concurrency@@YAIXZ() _GetConcurrency
-@ stub -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ
-@ stub -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ
+@ cdecl -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ() _StackGuard__GetCurrentInlineDepth
+@ cdecl -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ() _StackGuard__GetCurrentInlineDepth
 @ cdecl ?_GetNumberOfVirtualProcessors@_CurrentScheduler@details@Concurrency@@SAIXZ() _CurrentScheduler__GetNumberOfVirtualProcessors
 @ cdecl -arch=arm ?_GetScheduler@_Scheduler@details@Concurrency@@QAAPAVScheduler@3@XZ(ptr) _Scheduler__GetScheduler
 @ thiscall -arch=i386 ?_GetScheduler@_Scheduler@details@Concurrency@@QAEPAVScheduler@3@XZ(ptr) _Scheduler__GetScheduler
@@ -556,7 +556,7 @@
 @ cdecl -arch=arm ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAAKXZ(ptr) SpinWait__NumberOfSpins
 @ thiscall -arch=i386 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAEKXZ(ptr) SpinWait__NumberOfSpins
 @ cdecl -arch=win64 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IEAAKXZ(ptr) SpinWait__NumberOfSpins
-@ stub ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z
+@ cdecl ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z(long) Context_Oversubscribe
 @ cdecl -arch=arm ?_Reference@_Scheduler@details@Concurrency@@QAAIXZ(ptr) _Scheduler__Reference
 @ thiscall -arch=i386 ?_Reference@_Scheduler@details@Concurrency@@QAEIXZ(ptr) _Scheduler__Reference
 @ cdecl -arch=win64 ?_Reference@_Scheduler@details@Concurrency@@QEAAIXZ(ptr) _Scheduler__Reference
diff --git a/dlls/msvcr120_app/msvcr120_app.spec b/dlls/msvcr120_app/msvcr120_app.spec
index db7b344..6ed7918 100644
--- a/dlls/msvcr120_app/msvcr120_app.spec
+++ b/dlls/msvcr120_app/msvcr120_app.spec
@@ -18,9 +18,9 @@
 @ cdecl -arch=arm ??0_Condition_variable@details@Concurrency@@QAA@XZ(ptr) msvcr120.??0_Condition_variable@details@Concurrency@@QAA@XZ
 @ thiscall -arch=i386 ??0_Condition_variable@details@Concurrency@@QAE@XZ(ptr) msvcr120.??0_Condition_variable@details@Concurrency@@QAE@XZ
 @ cdecl -arch=win64 ??0_Condition_variable@details@Concurrency@@QEAA@XZ(ptr) msvcr120.??0_Condition_variable@details@Concurrency@@QEAA@XZ
-@ stub -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z
-@ stub -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z
-@ stub -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z
+@ cdecl -arch=arm ??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z(ptr ptr) msvcr120.??0_Context@details@Concurrency@@QAA@PAVContext@2@@Z
+@ thiscall -arch=i386 ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z(ptr ptr) msvcr120.??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z
+@ cdecl -arch=win64 ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z(ptr ptr) msvcr120.??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z
 @ cdecl -arch=arm ??0_Interruption_exception@details@Concurrency@@QAA@PBD@Z(ptr str) msvcr120.??0_Interruption_exception@details@Concurrency@@QAA@PBD@Z
 @ thiscall -arch=i386 ??0_Interruption_exception@details@Concurrency@@QAE@PBD@Z(ptr str) msvcr120.??0_Interruption_exception@details@Concurrency@@QAE@PBD@Z
 @ cdecl -arch=win64 ??0_Interruption_exception@details@Concurrency@@QEAA@PEBD@Z(ptr str) msvcr120.??0_Interruption_exception@details@Concurrency@@QEAA@PEBD@Z
@@ -375,9 +375,9 @@
 @ cdecl -arch=arm ??_F?$_SpinWait@$0A@@details@Concurrency@@QAAXXZ(ptr) msvcr120.??_F?$_SpinWait@$0A@@details@Concurrency@@QAAXXZ
 @ thiscall -arch=i386 ??_F?$_SpinWait@$0A@@details@Concurrency@@QAEXXZ(ptr) msvcr120.??_F?$_SpinWait@$0A@@details@Concurrency@@QAEXXZ
 @ cdecl -arch=win64 ??_F?$_SpinWait@$0A@@details@Concurrency@@QEAAXXZ(ptr) msvcr120.??_F?$_SpinWait@$0A@@details@Concurrency@@QEAAXXZ
-@ stub -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ
-@ stub -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ
-@ stub -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ
+@ cdecl -arch=arm ??_F_Context@details@Concurrency@@QAAXXZ(ptr) msvcr120.??_F_Context@details@Concurrency@@QAAXXZ
+@ thiscall -arch=i386 ??_F_Context@details@Concurrency@@QAEXXZ(ptr) msvcr120.??_F_Context@details@Concurrency@@QAEXXZ
+@ cdecl -arch=win64 ??_F_Context@details@Concurrency@@QEAAXXZ(ptr) msvcr120.??_F_Context@details@Concurrency@@QEAAXXZ
 @ cdecl -arch=arm ??_F_Scheduler@details@Concurrency@@QAAXXZ(ptr) msvcr120.??_F_Scheduler@details@Concurrency@@QAAXXZ
 @ thiscall -arch=i386 ??_F_Scheduler@details@Concurrency@@QAEXXZ(ptr) msvcr120.??_F_Scheduler@details@Concurrency@@QAEXXZ
 @ cdecl -arch=win64 ??_F_Scheduler@details@Concurrency@@QEAAXXZ(ptr) msvcr120.??_F_Scheduler@details@Concurrency@@QEAAXXZ
@@ -525,8 +525,8 @@
 @ stub -arch=win32 ?_GetConcRTTraceInfo@Concurrency@@YAPBU_CONCRT_TRACE_INFO@details@1@XZ
 @ stub -arch=win64 ?_GetConcRTTraceInfo@Concurrency@@YAPEBU_CONCRT_TRACE_INFO@details@1@XZ
 @ cdecl ?_GetConcurrency@details@Concurrency@@YAIXZ() msvcr120.?_GetConcurrency@details@Concurrency@@YAIXZ
-@ stub -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ
-@ stub -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ
+@ cdecl -arch=win32 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ() msvcr120.?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ
+@ cdecl -arch=win64 ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ() msvcr120.?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ
 @ cdecl ?_GetNumberOfVirtualProcessors@_CurrentScheduler@details@Concurrency@@SAIXZ() msvcr120.?_GetNumberOfVirtualProcessors@_CurrentScheduler@details@Concurrency@@SAIXZ
 @ cdecl -arch=arm ?_GetScheduler@_Scheduler@details@Concurrency@@QAAPAVScheduler@3@XZ(ptr) msvcr120.?_GetScheduler@_Scheduler@details@Concurrency@@QAAPAVScheduler@3@XZ
 @ thiscall -arch=i386 ?_GetScheduler@_Scheduler@details@Concurrency@@QAEPAVScheduler@3@XZ(ptr) msvcr120.?_GetScheduler@_Scheduler@details@Concurrency@@QAEPAVScheduler@3@XZ
@@ -553,7 +553,7 @@
 @ cdecl -arch=arm ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAAKXZ(ptr) msvcr120.?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAAKXZ
 @ thiscall -arch=i386 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAEKXZ(ptr) msvcr120.?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IAEKXZ
 @ cdecl -arch=win64 ?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IEAAKXZ(ptr) msvcr120.?_NumberOfSpins@?$_SpinWait@$0A@@details@Concurrency@@IEAAKXZ
-@ stub ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z
+@ cdecl ?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z(long) msvcr120.?_Oversubscribe@_Context@details@Concurrency@@SAX_N@Z
 @ cdecl -arch=arm ?_Reference@_Scheduler@details@Concurrency@@QAAIXZ(ptr) msvcr120.?_Reference@_Scheduler@details@Concurrency@@QAAIXZ
 @ thiscall -arch=i386 ?_Reference@_Scheduler@details@Concurrency@@QAEIXZ(ptr) msvcr120.?_Reference@_Scheduler@details@Concurrency@@QAEIXZ
 @ cdecl -arch=win64 ?_Reference@_Scheduler@details@Concurrency@@QEAAIXZ(ptr) msvcr120.?_Reference@_Scheduler@details@Concurrency@@QEAAIXZ
diff --git a/dlls/msvcrt/concurrency.c b/dlls/msvcrt/concurrency.c
index 02ead96..b8f5290 100644
--- a/dlls/msvcrt/concurrency.c
+++ b/dlls/msvcrt/concurrency.c
@@ -113,6 +113,7 @@ typedef struct {
     struct _StructuredTaskCollection *task_collection;
     CRITICAL_SECTION beacons_cs;
     struct list beacons;
+    size_t inline_depth;  /* _StackGuard::_GetCurrentInlineDepth */
 } ExternalContextBase;
 extern const vtable_ptr ExternalContextBase_vtable;
 static void ExternalContextBase_ctor(ExternalContextBase*);
@@ -1343,6 +1344,50 @@ BOOL __thiscall _Context_IsSynchronouslyBlocked(const _Context *this)
     TRACE("(%p)\n", this);
     return call_Context_IsSynchronouslyBlocked(this->context);
 }
+
+/* ??0_Context@details@Concurrency@@QAE@PAVContext@2@@Z */
+/* ??0_Context@details@Concurrency@@QEAA@PEAVContext@2@@Z */
+DEFINE_THISCALL_WRAPPER(_Context_ctor, 8)
+_Context* __thiscall _Context_ctor(_Context *this, Context *context)
+{
+    TRACE("(%p %p)\n", this, context);
+    this->context = context;
+    return this;
+}
+
+/* ??_F_Context@details@Concurrency@@QAEXXZ */
+/* ??_F_Context@details@Concurrency@@QEAAXXZ */
+/* the default constructor closure: no context */
+DEFINE_THISCALL_WRAPPER(_Context_default_ctor, 4)
+void __thiscall _Context_default_ctor(_Context *this)
+{
+#ifdef SG_MUTANT_CONTEXT_CLOSURE_CURRENT
+    this->context = Context_CurrentContext();
+    return;
+#endif
+    _Context_ctor(this, NULL);
+}
+
+/* ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAAIXZ */
+/* ?_GetCurrentInlineDepth@_StackGuard@details@Concurrency@@CAAEA_KXZ */
+/* the depth of chores the current context runs inline, which the
+ * _StackGuard objects of the headers count up and down */
+size_t* __cdecl _StackGuard__GetCurrentInlineDepth(void)
+{
+    static size_t unknown_depth;
+    ExternalContextBase *ctx = (ExternalContextBase*)get_current_context();
+
+    TRACE("()\n");
+
+    if (ctx->context.vtable != &ExternalContextBase_vtable) {
+        ERR("unknown context set\n");
+        return &unknown_depth;
+    }
+#ifdef SG_MUTANT_INLINE_DEPTH_SHARED
+    return &unknown_depth;
+#endif
+    return &ctx->inline_depth;
+}
 #endif
 
 DEFINE_THISCALL_WRAPPER(ExternalContextBase_GetId, 4)
